src/AbstractRecursiveIteratorLocator.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+namespace RebelCode\Modular\Locator;
+
+use Dhii\Exception\InvalidArgumentExceptionInterface;
+use Dhii\I18n\StringTranslatingTrait;
+use Dhii\I18n\StringTranslatorConsumingTrait;
+use Dhii\Modular\Locator\ModuleLocatorExceptionInterface;
+use Dhii\Validation\Exception\ValidationFailedExceptionInterface;
+use Exception;
+use Iterator;
+
+/**
+ * Abstract common functionality for locators that locate module configuration from recursive iterators or
+ * iterator iterators.
+ *
+ * @since [*next-version*]
+ */
+abstract class AbstractRecursiveIteratorLocator extends AbstractIteratorLocator
+{
+    /*
+     * Provides functionality for translating strings.
+     *
+     * @since [*next-version*]
+     */
+    use StringTranslatingTrait;
+
+    /**
+     * Reads a config source into a standard config.
+     *
+     * @since [*next-version*]
+     *
+     * @throws ModuleLocatorException      If an error occurred while reading the config.
+     * @throws CouldNotReadSourceException If the locator failed to read the config source.
+     *
+     * @return array The standard config.
+     */
+    protected function _read($config)
+    {
+        if (!($config instanceof Iterator) && !is_array($config)) {
+            throw $this->_createModuleLocatorException(
+                $this->__('Argument must be an array or iterator.'), null
```

Comment:
In a more common scenario, this would be an `InvalidArgumentException`. However, the consumer of the locator is unaware of this method and of what gets passed to it, and therefore a `ModuleLocatorExceptionInterface` is completely justified, since this is what the consumer knows may be thrown.

On the other hand, the message mentions that it's the "Argument" that is wrong. Given the above, I suggest that this is changed to read that "The config source must me an iterable". Note the word "iterable": perhaps, we can agree to use this [new PHP 7 type](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.iterable-pseudo-type) everywhere to mean e.g. `Iterator|array`.